### PR TITLE
Update Command Loop API for disconnects

### PR DESCRIPTION
**Pull request by muneebahmed10**
*Description*:

Clients are required to close network connections after sending a DISCONNECT packet. This updates the command loop API to return after a DISCONNECT packet is sent to allow the application to close the socket. In order to do so, the `MQTTAgent_CommandLoop` API is updated to return an `MQTTStatus_t`, and take two pointer parameters for an `MQTTContextHandle_t` and `MQTTContext_t`:
 * The return value will be `MQTTSuccess` for deliberate disconnects and command loop terminations, and an appropriate error code for errors.
 * The context handle will be set to the appropriate handle for disconnects and errors, and will not be set for termination.
 * The context pointer will be set to the context corresponding to the handle, or `NULL` for termination. This is necessary due to the fact that `pMqttContext->connectStatus` must be set to `MQTTNotConnected` by the agent task on errors, as the agent code requires this to be set if attempting an MQTT connection. This parameter may be able to be removed if an API function is added to retrieve contexts from handles, or the corresponding check is removed from `MQTTAgent_Connect`.
 * So, disconnects and terminations can be differentiated from errors by the return status, and can be differentiated from each other by testing the context pointer for `NULL`.

While making these changes, a bug was also fixed which would always return the last MQTT context in the array instead of the one which caused an error.

In addition to the changes required to handle disconnects, this adds the `backoffAlgorithm` submodule, and makes some minor changes for consistency in style. Each change is a separate commit.

### CODE REVIEW

File: lib/FreeRTOS/freertos-plus-mqtt/freertos_mqtt_agent.c [hunk 1]
```diff
@@ -1432,43 +1444,57 @@ MQTTStatus_t MQTTAgent_Init( MQTTContextHandle_t mqttContextHandle,
...
     configASSERT( commandQueue );
 
+    if( ( pOutContextHandle == NULL ) || ( pOutContext == NULL ) )
+    {
+        operationStatus = MQTTBadParameter;
+    }
+
     /* Loop until we receive a terminate command. */
-    for( ; ; )
+    for( ; operationStatus == MQTTSuccess; )
```
Comment (yourslab) [inline lib/FreeRTOS/freertos-plus-mqtt/freertos_mqtt_agent.c:1463]: Would a while loop be more appropriate here?

File: lib/FreeRTOS/freertos-plus-mqtt/freertos_mqtt_agent.c [hunk 2]
```diff
@@ -1387,11 +1396,14 @@ MQTTStatus_t MQTTAgent_Init( MQTTContextHandle_t mqttContextHandle,
...
-                                       sizeof( Command_t * ),
-                                       staticQueueStorageArea,
-                                       &staticQueueStructure );
+    if( commandQueue == NULL )  /*_RB_ Needs to change so two contexts can be used in the same agent. */
+    {
+        LogDebug( ( "Creating command queue." ) );
+        commandQueue = xQueueCreateStatic( MQTT_AGENT_COMMAND_QUEUE_LENGTH,
+                                           sizeof( Command_t * ),
+                                           staticQueueStorageArea,
+                                           &staticQueueStructure );
```
Comment (yourslab) [inline lib/FreeRTOS/freertos-plus-mqtt/freertos_mqtt_agent.c:1405]: Should users have the ability to allocate this queue from the heap?

File: lib/FreeRTOS/freertos-plus-mqtt/freertos_mqtt_agent.c [hunk 3]
```diff
@@ -1432,43 +1444,57 @@ MQTTStatus_t MQTTAgent_Init( MQTTContextHandle_t mqttContextHandle,
...
-        /* Return the current MQTT context if status was not successful. */
-        if( operationStatus != MQTTSuccess )
+        /* Return the current MQTT context on disconnect or error. */
+        if( ( currentCommandType == DISCONNECT ) || ( operationStatus != MQTTSuccess ) )
         {
-            LogError( ( "MQTT operation failed with status %s\n",
-                        MQTT_Status_strerror( operationStatus ) ) );
+            /* Retrieve context handle from context. */
+            configASSERT( ( *pOutContext >= mqttContexts ) &&
+                          ( *pOutContext <= &mqttContexts[ MQTT_AGENT_MAX_SIMULTANEOUS_CONNECTIONS - 1 ] ) );
```
Comment (yourslab) [inline lib/FreeRTOS/freertos-plus-mqtt/freertos_mqtt_agent.c:1477]: *pOutContext is of type `MQTTContext_t *` but `&mqttContexts[ MQTT_AGENT_MAX_SIMULTANEOUS_CONNECTIONS - 1 ]` is a double pointer right? Should the `&` be removed?

File: lib/FreeRTOS/freertos-plus-mqtt/freertos_mqtt_agent.c [hunk 4]
```diff
@@ -1797,17 +1822,17 @@ MQTTStatus_t MQTTAgent_Free( MQTTContextHandle_t mqttContextHandle,
...
 
 /*-----------------------------------------------------------*/
 
-MQTTStatus_t MQTTAgent_Terminate( uint32_t blockTimeMS )
+MQTTStatus_t MQTTAgent_Terminate( uint32_t blockTimeMs )
 {
     MQTTStatus_t statusReturn = MQTTSuccess;
 
-    statusReturn = createAndAddCommand( TERMINATE, 0, NULL, NULL, NULL, NULL, NULL, blockTimeMS );
+    statusReturn = createAndAddCommand( TERMINATE, 0, NULL, NULL, NULL, NULL, NULL, blockTimeMs );
```
Comment (yourslab) [inline lib/FreeRTOS/freertos-plus-mqtt/freertos_mqtt_agent.c:1835]: IMO this is a little unreadable. I think that technique you had of using a struct to make it clear what parameter is being assigned would come in handy here.

File: lib/FreeRTOS/freertos-plus-mqtt/freertos_mqtt_agent.c (hunk 3 above)
Comment (muneebahmed10) [reply lib/FreeRTOS/freertos-plus-mqtt/freertos_mqtt_agent.c:1477]: `mqttContexts` is an array of `MQTTContext_t`. The `[]` gives the last element, and the `&` returns its address, so it's only a single pointer. `(mqttContexts + MQTT_AGENT_MAX_SIMULTANEOUS_CONNECTIONS - 1)` is what we want

File: lib/FreeRTOS/freertos-plus-mqtt/freertos_mqtt_agent.c (hunk 1 above)
Comment (muneebahmed10) [reply lib/FreeRTOS/freertos-plus-mqtt/freertos_mqtt_agent.c:1463]: Probably

File: lib/FreeRTOS/freertos-plus-mqtt/freertos_mqtt_agent.c (hunk 2 above)
Comment (muneebahmed10) [reply lib/FreeRTOS/freertos-plus-mqtt/freertos_mqtt_agent.c:1405]: That will be part of the queue abstraction, which we haven't designed yet. This change was necessary here to allow multiple connections though

File: lib/FreeRTOS/freertos-plus-mqtt/freertos_mqtt_agent.c (hunk 4 above)
Comment (muneebahmed10) [reply lib/FreeRTOS/freertos-plus-mqtt/freertos_mqtt_agent.c:1835]: Possibly, however this actually is the function that validates and assigns these parameters to the struct, and the process of "Assign to struct -> validate params -> get new struct from pool -> copy to retrieved struct" just seems wrong. "Get new struct from pool -> Assign to struct -> validate params" could work, but it still seems weird to validate after the assignment.

File: source/connection_manager.c [hunk 5]
```diff
@@ -587,22 +614,31 @@ static void prvMQTTAgentTask( void * pvParameters )
...
 
-        /* Context is only returned if error occurred which will may have
-         * disconnected the socket from the MQTT broker already. */
-        if( pMqttContext != NULL )
+        /* Context is returned for disconnects or errors. The socket should
+         * be disconnected. */
+        if( ( xMQTTStatus == MQTTSuccess ) && ( pMqttContext != NULL ) )
+        {
+            /* MQTT Disconnect. Disconnect the socket. */
+            xNetworkResult = prvSocketDisconnect( &xNetworkContext );
```
Comment (leegeth) [inline source/connection_manager.c:636]: Is there only one NetworkContext for different MQTTContexts?

File: source/connection_manager.c [hunk 6]
```diff
@@ -587,22 +614,31 @@ static void prvMQTTAgentTask( void * pvParameters )
...
+        }
```
Comment (leegeth) [inline source/connection_manager.c:637]: Does the do while loop need to be terminated on Disconnect?

File: source/connection_manager.c [hunk 7]
```diff
@@ -587,22 +614,31 @@ static void prvMQTTAgentTask( void * pvParameters )
...
+        /* Error. */
+        else if( pMqttContext != NULL )
```
Comment (leegeth) [inline source/connection_manager.c:639]: So this loop exits if there is pMqttContext is NULL. When will that happen? on a TERMINATE command?

**Pull request closed without merging**